### PR TITLE
[TIMOB-19486] Android: Generate missing app icons based on DefaultIcon.png

### PR DESCRIPTION
**Pull request by feons**
https://jira.appcelerator.org/browse/TIMOB-19486

### CODE REVIEW

File: android/cli/commands/_build.js [hunk 1]
```diff
@@ -2510,6 +2510,113 @@ AndroidBuilder.prototype.copyResources = function copyResources(next) {
...
+				defaultIcon = icon;
+				return true;
+			}
+		});
+
+		if (defaultIcon) {
+			var currDefaultIconHash = this.hash(fs.readFileSync(defaultIcon)),
+				defaultIconRelative = path.relative(this.projectDir, defaultIcon),
+				prevDefaultIcon = previousBuildManifestFiles && previousBuildManifestFiles[defaultIconRelative],
+				prevDefaultIconHash = prevDefaultIcon && prevDefaultIcon.hash,
```
Comment (cb1kenobi) [inline android/cli/commands/_build.js]: Comma should be a semi-colon.

File: android/cli/commands/_build.js [hunk 2]
```diff
@@ -2510,6 +2510,113 @@ AndroidBuilder.prototype.copyResources = function copyResources(next) {
...
+
+			this.currentBuildManifestFiles[defaultIconRelative] = {
+				hash: currDefaultIconHash
+			};
+
+			defaultIconChanged = prevDefaultIconHash !== currDefaultIconHash;
+		}
+
+		function copyGenAppIcons() {
```
Comment (cb1kenobi) [inline android/cli/commands/_build.js]: This is sort of confusing. Maybe we just inline this function in the `this.generateAppIcons()` call?

File: android/cli/commands/_build.js [hunk 3]
```diff
@@ -2510,6 +2510,113 @@ AndroidBuilder.prototype.copyResources = function copyResources(next) {
 		}
 	}, this);
 
+	// genereate missing app icons
+	tasks.push(function (cb) {
+		this.currentBuildManifestFiles = this.currentBuildManifestFiles || {};
```
Comment (cb1kenobi) [inline android/cli/commands/_build.js]: Android build doesn't use differential builds... yet. There is no need for `this.currentBuildManifestFiles`. All references should be removed.

File: android/cli/commands/_build.js [hunk 4]
```diff
@@ -2510,6 +2510,113 @@ AndroidBuilder.prototype.copyResources = function copyResources(next) {
...
+			// the app icon was previously resized
+			if (!defaultIconChanged && fs.existsSync(dest)) {
+				this.logger.trace(__('Found %sx%s app icon: %s', icon.width, icon.height, dest.cyan));
+				return;
+			}
+
+			// generated icons will be placed under ../build/android/bin/assets/res/drawable-[density]
+			fs.existsSync(path.dirname(dest)) || wrench.mkdirSyncRecursive(path.dirname(dest));
+			missingIcons.push({
+				description: __('%s', description),
```
Comment (cb1kenobi) [inline android/cli/commands/_build.js]: Not sure why you can't just set `description: description,`.

File: android/cli/commands/_build.js (hunk 3 above)
Comment (feons) [inline android/cli/commands/_build.js]: That would mean, we will regenerate the app icons on every build/rebuild?

File: android/cli/commands/_build.js (hunk 4 above)
Comment (feons) [inline android/cli/commands/_build.js]: me neither :s

File: android/cli/commands/_build.js (hunk 3 above)
Comment (cb1kenobi) [inline android/cli/commands/_build.js]: Part of the problem is `this.currentBuildManifestFiles` is something you created. It doesn't exist before this PR. It is never persisted to disk. So, it has no effect.

The way I see it, you can either remove `this.currentBuildManifestFiles` in which case the missing icons will be generated every build -or- hack something to track this state.

Simply checking if the destination image exists would be a step in the right direction. Combine that with checking the mtime on the source DefaultIcon*.png with the mtime of the generated images and you should have an OK mechanism.

**Comment by cheekiatng** [comment]
APPROVED

**Comment by cheekiatng** [comment]
Sorry I thought it was addressed. Reverting PR. I would prefer checking the mtime mechanism.

**Comment by cheekiatng** [comment]
@feons can you create another PR that addresses that issue? I'll reopen the JIRA ticket.

File: android/cli/commands/_build.js [hunk 5]
```diff
@@ -2510,6 +2510,120 @@ AndroidBuilder.prototype.copyResources = function copyResources(next) {
...
+		this.currentBuildManifestFiles = this.currentBuildManifestFiles || {};
+
+		var lookup = {
+				// ../Resources/android/appicon.png
+				'default':	{ height: 128, width: 128, required: true },
+				// ../platform/android/res/drawable-[key]/appicon.png
+				'ldpi':		{ height: 36, width: 36, required: false },
+				'mdpi':		{ height: 48, width: 48, required: false },
+				'hdpi':		{ height: 72, width: 72, required: false },
+				'xhdpi':	{ height: 96, width: 96, required: false },
```
Comment (cb1kenobi) [inline android/cli/commands/_build.js]: Missing padding between `:` and value.

**Pull request merged**